types/block.go

```diff
@@ -65,46 +66,46 @@ func (b *Block) ValidateBasic() error {
 	defer b.mtx.Unlock()
 
 	if b.Height < 0 {
-		return fmt.Errorf(
-			"Negative Block.Header.Height: %v",
-			b.Height,
-		)
+		return fmt.Errorf("Negative Block.Header.Height: %v", b.Height)
 	}
 
 	newTxs := int64(len(b.Data.Txs))
 	if b.NumTxs != newTxs {
-		return fmt.Errorf(
-			"Wrong Block.Header.NumTxs. Expected %v, got %v",
+		return fmt.Errorf("Wrong Block.Header.NumTxs. Expected %v, got %v",
 			newTxs,
 			b.NumTxs,
 		)
 	}
+
 	if !bytes.Equal(b.LastCommitHash, b.LastCommit.Hash()) {
-		return fmt.Errorf(
-			"Wrong Block.Header.LastCommitHash.  Expected %v, got %v",
+		return fmt.Errorf("Wrong Block.Header.LastCommitHash. Expected %v, got %v",
 			b.LastCommitHash,
 			b.LastCommit.Hash(),
 		)
 	}
+
 	if b.Header.Height != 1 {
 		if err := b.LastCommit.ValidateBasic(); err != nil {
 			return err
 		}
 	}
+
 	if !bytes.Equal(b.DataHash, b.Data.Hash()) {
 		return fmt.Errorf(
 			"Wrong Block.Header.DataHash.  Expected %v, got %v",
 			b.DataHash,
 			b.Data.Hash(),
 		)
 	}
+
 	if !bytes.Equal(b.EvidenceHash, b.Evidence.Hash()) {
 		return fmt.Errorf(
 			"Wrong Block.Header.EvidenceHash.  Expected %v, got %v",
 			b.EvidenceHash,
 			b.Evidence.Hash(),
 		)
 	}
+
 	return nil
```

Comment:
We should also do `VerifyBasic` on the evidence here since otherwise the votes in the `DuplicateVoteEvidence` could still be `nil` when the block is further processed.

Reply:
```suggestion
	for _, ev := range b.Evidence.Evidence {
		if err := ev.ValidateBasic(); err != nil {
			return err
		}
	}
```

Reply:
@ebuchman @melekes This would be a quick fix. However `evidence` is of the type `EvidenceData` which should also have a `ValidateBasic` function. That would also have to check whether `EvidenceData.Evidence != nil` since a nil slice could be passed which could cause another `Consensus Failure`.

We have to check this before doing `b.Evidence.Hash()` in the lines above since it could otherwise crash there!

This sounds like another bounty or do we count this as included in the first finding ? ;)

Reply:
`b.Evidence` can't be nil since it's not a pointer. `b.Evidence.Evidence` could be nil because it's a slice,  but here we just loop over it, so if its nil its no problem, the loop just won't run. The methods defined on EvidenceList similarly only loop over it, so if it's nil there should be no issue. Am I missing something? Or you're saying we should always ensure we have a zero-length array here and not nil ?